File: frontend/appflowy_tauri/src/appflowy_app/stores/effects/database/field/type_option/type_option_context.ts
import { None, Ok, Option, Result, Some } from 'ts-results';
import { TypeOptionController } from './type_option_controller';
import {
  CheckboxTypeOptionPB,
  ChecklistTypeOptionPB,
  DateTypeOptionPB,
  FlowyError,
  MultiSelectTypeOptionPB,
  NumberTypeOptionPB,
  SingleSelectTypeOptionPB,
  URLTypeOptionPB,
} from '@/services/backend';
import { utf8Decoder, utf8Encoder } from '../../cell/data_parser';
import { DatabaseFieldObserver } from '../field_observer';

abstract class TypeOptionSerde<T> {
  abstract deserialize(buffer: Uint8Array): T;

  abstract serialize(value: T): Uint8Array;
}

// RichText
export function makeRichTextTypeOptionContext(controller: TypeOptionController): RichTextTypeOptionContext {
  const parser = new RichTextTypeOptionSerde();
  return new TypeOptionContext<string>(parser, controller);
}

export type RichTextTypeOptionContext = TypeOptionContext<string>;

class RichTextTypeOptionSerde extends TypeOptionSerde<string> {
  deserialize(buffer: Uint8Array): string {
    return utf8Decoder.decode(buffer);
  }

  serialize(value: string): Uint8Array {
    return utf8Encoder.encode(value);
  }
}

// Number
export function makeNumberTypeOptionContext(controller: TypeOptionController): NumberTypeOptionContext {
  const parser = new NumberTypeOptionSerde();
  return new TypeOptionContext<NumberTypeOptionPB>(parser, controller);
}

export type NumberTypeOptionContext = TypeOptionContext<NumberTypeOptionPB>;

class NumberTypeOptionSerde extends TypeOptionSerde<NumberTypeOptionPB> {
  deserialize(buffer: Uint8Array): NumberTypeOptionPB {
    return NumberTypeOptionPB.deserializeBinary(buffer);
  }

  serialize(value: NumberTypeOptionPB): Uint8Array {
    return value.serializeBinary();
  }
}

// Checkbox
export function makeCheckboxTypeOptionContext(controller: TypeOptionController): CheckboxTypeOptionContext {
  const parser = new CheckboxTypeOptionSerde();
  return new TypeOptionContext<CheckboxTypeOptionPB>(parser, controller);
}

export type CheckboxTypeOptionContext = TypeOptionContext<CheckboxTypeOptionPB>;

class CheckboxTypeOptionSerde extends TypeOptionSerde<CheckboxTypeOptionPB> {
  deserialize(buffer: Uint8Array): CheckboxTypeOptionPB {
    return CheckboxTypeOptionPB.deserializeBinary(buffer);
  }

  serialize(value: CheckboxTypeOptionPB): Uint8Array {
    return value.serializeBinary();
  }
}

// URL
export function makeURLTypeOptionContext(controller: TypeOptionController): URLTypeOptionContext {
  const parser = new URLTypeOptionSerde();
  return new TypeOptionContext<URLTypeOptionPB>(parser, controller);
}

export type URLTypeOptionContext = TypeOptionContext<URLTypeOptionPB>;

class URLTypeOptionSerde extends TypeOptionSerde<URLTypeOptionPB> {
  deserialize(buffer: Uint8Array): URLTypeOptionPB {
    return URLTypeOptionPB.deserializeBinary(buffer);
  }

  serialize(value: URLTypeOptionPB): Uint8Array {
    return value.serializeBinary();
  }
}

// Date
export function makeDateTypeOptionContext(controller: TypeOptionController): DateTypeOptionContext {
  const parser = new DateTypeOptionSerde();
  return new TypeOptionContext<DateTypeOptionPB>(parser, controller);
}

export type DateTypeOptionContext = TypeOptionContext<DateTypeOptionPB>;

class DateTypeOptionSerde extends TypeOptionSerde<DateTypeOptionPB> {
  deserialize(buffer: Uint8Array): DateTypeOptionPB {
    return DateTypeOptionPB.deserializeBinary(buffer);
  }

  serialize(value: DateTypeOptionPB): Uint8Array {
    return value.serializeBinary();
  }
}

// SingleSelect
export function makeSingleSelectTypeOptionContext(controller: TypeOptionController): SingleSelectTypeOptionContext {
  const parser = new SingleSelectTypeOptionSerde();
  return new TypeOptionContext<SingleSelectTypeOptionPB>(parser, controller);
}

export type SingleSelectTypeOptionContext = TypeOptionContext<SingleSelectTypeOptionPB>;

class SingleSelectTypeOptionSerde extends TypeOptionSerde<SingleSelectTypeOptionPB> {
  deserialize(buffer: Uint8Array): SingleSelectTypeOptionPB {
    return SingleSelectTypeOptionPB.deserializeBinary(buffer);
  }

  serialize(value: SingleSelectTypeOptionPB): Uint8Array {
    return value.serializeBinary();
  }
}

// Multi-select
export function makeMultiSelectTypeOptionContext(controller: TypeOptionController): MultiSelectTypeOptionContext {
  const parser = new MultiSelectTypeOptionSerde();
  return new TypeOptionContext<MultiSelectTypeOptionPB>(parser, controller);
}

export type MultiSelectTypeOptionContext = TypeOptionContext<MultiSelectTypeOptionPB>;

class MultiSelectTypeOptionSerde extends TypeOptionSerde<MultiSelectTypeOptionPB> {
  deserialize(buffer: Uint8Array): MultiSelectTypeOptionPB {
    return MultiSelectTypeOptionPB.deserializeBinary(buffer);
  }

  serialize(value: MultiSelectTypeOptionPB): Uint8Array {
    return value.serializeBinary();
  }
}

// Checklist
export function makeChecklistTypeOptionContext(controller: TypeOptionController): ChecklistTypeOptionContext {
  const parser = new ChecklistTypeOptionSerde();
  return new TypeOptionContext<ChecklistTypeOptionPB>(parser, controller);
}

export type ChecklistTypeOptionContext = TypeOptionContext<ChecklistTypeOptionPB>;

class ChecklistTypeOptionSerde extends TypeOptionSerde<ChecklistTypeOptionPB> {
  deserialize(buffer: Uint8Array): ChecklistTypeOptionPB {
    return ChecklistTypeOptionPB.deserializeBinary(buffer);
  }

  serialize(value: ChecklistTypeOptionPB): Uint8Array {
    return value.serializeBinary();
  }
}

export class TypeOptionContext<T> {
  private typeOption: Option<T>;
  private fieldObserver: DatabaseFieldObserver;

  constructor(public readonly parser: TypeOptionSerde<T>, private readonly controller: TypeOptionController) {
    this.typeOption = None;
    this.fieldObserver = new DatabaseFieldObserver(controller.fieldId);

    void this.fieldObserver.subscribe({
      onFieldChanged: () => {
        void this.getTypeOption();
      },
    });
  }

  get viewId(): string {
    return this.controller.viewId;
  }

  getTypeOption = async (): Promise<Result<T, FlowyError>> => {
    const result = await this.controller.getTypeOption();
    if (result.ok) {
      const typeOption = this.parser.deserialize(result.val.type_option_data);
      this.typeOption = Some(typeOption);
      return Ok(typeOption);
    } else {
      return result;
    }
  };

  // Save the typeOption to disk
  setTypeOption = async (typeOption: T) => {
    await this.controller.saveTypeOption(this.parser.serialize(typeOption));
    this.typeOption = Some(typeOption);
  };

  dispose = async () => {
    await this.fieldObserver.unsubscribe();
  };
}

File: frontend/appflowy_flutter/lib/plugins/database_view/application/field/type_option/type_option_context.dart
import 'package:appflowy_backend/dispatch/dispatch.dart';
import 'package:appflowy_backend/protobuf/flowy-database2/checkbox_entities.pb.dart';
import 'package:appflowy_backend/protobuf/flowy-database2/date_entities.pb.dart';
import 'package:appflowy_backend/protobuf/flowy-database2/number_entities.pb.dart';
import 'package:appflowy_backend/protobuf/flowy-database2/select_option.pb.dart';
import 'package:appflowy_backend/protobuf/flowy-database2/text_entities.pb.dart';
import 'package:appflowy_backend/protobuf/flowy-database2/url_entities.pb.dart';
import 'package:appflowy_backend/protobuf/flowy-error/errors.pb.dart';
import 'package:appflowy_backend/protobuf/flowy-database2/field_entities.pb.dart';
import 'package:dartz/dartz.dart';
import 'package:protobuf/protobuf.dart';
import 'type_option_data_controller.dart';

abstract class TypeOptionParser<T> {
  T fromBuffer(List<int> buffer);
}

// Number
typedef NumberTypeOptionContext = TypeOptionContext<NumberTypeOptionPB>;

class NumberTypeOptionWidgetDataParser
    extends TypeOptionParser<NumberTypeOptionPB> {
  @override
  NumberTypeOptionPB fromBuffer(List<int> buffer) {
    return NumberTypeOptionPB.fromBuffer(buffer);
  }
}

// RichText
typedef RichTextTypeOptionContext = TypeOptionContext<RichTextTypeOptionPB>;

class RichTextTypeOptionWidgetDataParser
    extends TypeOptionParser<RichTextTypeOptionPB> {
  @override
  RichTextTypeOptionPB fromBuffer(List<int> buffer) {
    return RichTextTypeOptionPB.fromBuffer(buffer);
  }
}

// Checkbox
typedef CheckboxTypeOptionContext = TypeOptionContext<CheckboxTypeOptionPB>;

class CheckboxTypeOptionWidgetDataParser
    extends TypeOptionParser<CheckboxTypeOptionPB> {
  @override
  CheckboxTypeOptionPB fromBuffer(List<int> buffer) {
    return CheckboxTypeOptionPB.fromBuffer(buffer);
  }
}

// URL
typedef URLTypeOptionContext = TypeOptionContext<URLTypeOptionPB>;

class URLTypeOptionWidgetDataParser extends TypeOptionParser<URLTypeOptionPB> {
  @override
  URLTypeOptionPB fromBuffer(List<int> buffer) {
    return URLTypeOptionPB.fromBuffer(buffer);
  }
}

// Date
typedef DateTypeOptionContext = TypeOptionContext<DateTypeOptionPB>;

class DateTypeOptionDataParser extends TypeOptionParser<DateTypeOptionPB> {
  @override
  DateTypeOptionPB fromBuffer(List<int> buffer) {
    return DateTypeOptionPB.fromBuffer(buffer);
  }
}

// SingleSelect
typedef SingleSelectTypeOptionContext
    = TypeOptionContext<SingleSelectTypeOptionPB>;

class SingleSelectTypeOptionWidgetDataParser
    extends TypeOptionParser<SingleSelectTypeOptionPB> {
  @override
  SingleSelectTypeOptionPB fromBuffer(List<int> buffer) {
    return SingleSelectTypeOptionPB.fromBuffer(buffer);
  }
}

// Multi-select
typedef MultiSelectTypeOptionContext
    = TypeOptionContext<MultiSelectTypeOptionPB>;

class MultiSelectTypeOptionWidgetDataParser
    extends TypeOptionParser<MultiSelectTypeOptionPB> {
  @override
  MultiSelectTypeOptionPB fromBuffer(List<int> buffer) {
    return MultiSelectTypeOptionPB.fromBuffer(buffer);
  }
}

// Multi-select
typedef ChecklistTypeOptionContext = TypeOptionContext<ChecklistTypeOptionPB>;

class ChecklistTypeOptionWidgetDataParser
    extends TypeOptionParser<ChecklistTypeOptionPB> {
  @override
  ChecklistTypeOptionPB fromBuffer(List<int> buffer) {
    return ChecklistTypeOptionPB.fromBuffer(buffer);
  }
}

class TypeOptionContext<T extends GeneratedMessage> {
  T? _typeOptionObject;
  final TypeOptionParser<T> dataParser;
  final TypeOptionController _dataController;

  TypeOptionContext({
    required this.dataParser,
    required TypeOptionController dataController,
  }) : _dataController = dataController;

  String get viewId => _dataController.loader.viewId;

  String get fieldId => _dataController.field.id;

  Future<T> loadTypeOptionData({
    void Function(T)? onCompleted,
    required void Function(FlowyError) onError,
  }) async {
    await _dataController.reloadTypeOption().then((result) {
      result.fold((l) => null, (err) => onError(err));
    });

    onCompleted?.call(typeOption);
    return typeOption;
  }

  T get typeOption {
    if (_typeOptionObject != null) {
      return _typeOptionObject!;
    }

    final T object = _dataController.getTypeOption(dataParser);
    _typeOptionObject = object;
    return object;
  }

  set typeOption(T typeOption) {
    _dataController.typeOptionData = typeOption.writeToBuffer();
    _typeOptionObject = typeOption;
  }
}

abstract class TypeOptionFieldDelegate {
  void onFieldChanged(void Function(String) callback);
  void dispose();
}

abstract class ITypeOptionLoader {
  String get viewId;
  String get fieldName;

  Future<Either<TypeOptionPB, FlowyError>> initialize();
}

/// Uses when editing a existing field
class FieldTypeOptionLoader {
  final String viewId;
  final FieldPB field;

  FieldTypeOptionLoader({
    required this.viewId,
    required this.field,
  });

  Future<Either<TypeOptionPB, FlowyError>> load() {
    final payload = TypeOptionPathPB.create()
      ..viewId = viewId
      ..fieldId = field.id
      ..fieldType = field.fieldType;

    return DatabaseEventGetTypeOption(payload).send();
  }
}
